libraries/ms-common/src/stm32f0xx/interrupt.c

```diff
@@ -0,0 +1,200 @@
+#include "interrupt.h"
+
+#include <stdint.h>
+#include <stdlib.h>
+
+#include "gpio.h"
+#include "gpio_cfg.h"
+#include "status.h"
+#include "stm32f0xx_exti.h"
+#include "stm32f0xx_misc.h"
+#include "stm32f0xx_rcc.h"
+#include "stm32f0xx_syscfg.h"
+
+#define NUM_INTERRUPT_GPIO_CHANNELS 3
+static InterruptPriority s_interrupt_gpio_priority_map[NUM_INTERRUPT_GPIO_CHANNELS];
+
+void interrupt_init() {
+  RCC_APB2PeriphClockCmd(RCC_APB2Periph_SYSCFG, ENABLE);
+
+  for (int i = 0; i < NUM_INTERRUPT_GPIO_CHANNELS; i++) {
+    s_interrupt_gpio_priority_map[i] = NUM_INTERRUPT_PRIORITY;
+  }
+}
+
+#pragma GCC push_options
+#pragma GCC optimize("O0")
+// Use this to count how many nested disables there are and only enable once out of all critical
+// sections.
+static uint8_t s_interrupt_priority_mask = 0;
+
+void interrupt_enable() {
+  if (s_interrupt_priority_mask > 0) {
```

Comment:
The paradigm I've seen most often is a pretty simple "bool represents whether interrupts were disabled before". I think it might be simpler and a bit more robust.

i.e. interrupt_disable returns whether interrupts were enabled, and interrupt_enable takes the bool from interrupt_disable and acts accordingly. The goal is to always preserve state.